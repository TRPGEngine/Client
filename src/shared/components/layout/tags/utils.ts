import { Fragment } from 'react';
import _get from 'lodash/get';
import _set from 'lodash/set';
import _isNil from 'lodash/isNil';
import _clone from 'lodash/clone';
import _toPairs from 'lodash/toPairs';
import _fromPairs from 'lodash/fromPairs';
import type {
  XMLBuilderState,
  XMLBuilderContext,
} from '@shared/components/layout/XMLBuilder';
import {
  StateDataType,
  StateActionType,
} from '@shared/components/layout/types';
import type { XMLElement } from '../parser/xml-parser';

export type OperationDataType = {
  scope: string;
  field: string;
};

/**
 * 获取需要操作的变量的作用域与操作的变量名
 * 作用域默认为data
 * 用于独立一个path的第一段
 * example:
 * - 'a' => {scope: 'data', field: 'a'}
 * - 'a.b' => {scope: 'a', field: 'b'}
 * - 'a.b.c' => {scope: 'a', field: 'b.c'}
 *
 * @param str 操作参数字符串
 */
export const getOperationData = (str: string): OperationDataType => {
  const [scope, ...fields] = str.split('.');

  if (fields.length > 0) {
    // 如果为abc.def
    return {
      scope,
      field: fields.join('.'),
    };
  } else {
    // 如果为abc
    return {
      scope: 'data',
      field: scope,
    };
  }
};

/**
 * 尝试将文本转化为数字
 * 用于表单数据转换
 * @param str 要转换的文本
 */
export const tryToNumber = (str: string): number | string => {
  if (typeof str === 'string' && str === '') {
    // 空字符串直接返回空字符串
    return '';
  }

  if (typeof str === 'string' && (str.length > 15 || str.startsWith(' '))) {
    // 大于15位的数字视为字符串
    // 输入空格视为字符串
    return str;
  }

  const num = Number(str);
  return !isNaN(num) ? num : str;
};

/**
 * 标准化标签名。返回字符串或空节点
 * @param tagName 标签名
 */
export const normalizeTagName = (tagName: string): string | typeof Fragment => {
  // 如果是首字母开头。视为没有做定义的内置操作。改为t-xxx的格式防止抛出命名警告
  if (typeof tagName === 'string' && /[A-Z]/.test(tagName[0])) {
    tagName = 't-' + tagName.toLowerCase();
  }

  // 如果是空字符串或者undefined。使用react的Fragment
  if (!tagName) {
    return Fragment;
  }

  return tagName;
};

/**
 * 获取要实际操作变量名的路径
 *
 * 返回要获取变量在data数据中的位置
 */
export const getNamePath = (name: string) => {
  const { scope, field } = getOperationData(name);
  let path = field;
  if (scope !== 'data') {
    path = [scope, field].join('.');
  }

  return path;
};

/**
 * 根据上下文获取指定状态的数据
 */
export const getStateValue = (
  context: XMLBuilderContext,
  bindingName: string
): StateDataType => {
  const path = getNamePath(bindingName);

  let ret = _get(context.state.data, path);
  if (_isNil(ret)) {
    // 如果没有从data中取到。可能path是有前缀的路径。尝试从上一级获取
    ret = _get(context.state, path);
  }

  // NOTICE: 不知道为什么要不允许获取object
  // 直接返回匹配的数据，如果有问题再改回来
  // if (typeof ret === 'object') {
  //   return '';
  // }

  return ret as any;
};

/**
 * 设置状态的值
 */
export const setStateValue = (
  context: XMLBuilderContext,
  bindingName: string,
  value: StateDataType
): void => {
  const { dispatch } = context;
  const { scope, field } = getOperationData(bindingName);

  dispatch({
    type: StateActionType.UpdateData,
    payload: {
      scope,
      field,
      value,
    },
  });
};
/**
 * 将属性中额私有参数提取出来(以下划线开头_的成员变量)
 * 返回一个新的对象
 */
export const removePrivateProps = <T extends object>(props: T): Partial<T> => {
  return _fromPairs(
    _toPairs(props).filter(([key, value]) => !key.startsWith('_'))
  ) as T;
};

/**
 * 处理xml读取到的多行文本
 */
export const parseMultilineText = (text: string) => {
  // 支持\n的渲染 拿到的换行符为\\n
  if (_isNil(text) || typeof text !== 'string') {
    return '';
  }

  return text.replace(new RegExp('\\\\n', 'g'), '\n');
};

/**
 * 获取元素子节点的文本
 */
export const getChildrenText = (el: XMLElement): string => {
  const childrenEl = el.elements!;

  return childrenEl
    .map<string>((el) => {
      if (el.type === 'text') {
        return String(el.text);
      } else if (el.type === 'element') {
        return getChildrenText(el);
      } else {
        return '';
      }
    })
    .reduce((prev, curr) => {
      return prev + curr;
    }, '');
};
