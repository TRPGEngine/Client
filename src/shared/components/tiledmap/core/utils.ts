import type { Size, Position, Rect } from './types';

/**
 * 计算距离数最近的网格点
 * @param num 真实位置
 * @param gridSizeNum 网格大小
 */
function calcPx2GridMagnetNum(num: number, gridSizeNum: number): number {
  return Math.round(num / gridSizeNum);
}

/**
 * 计算距离数最左上角的点
 * @param num
 * @param gridSizeNum
 */
function calcPx2GridFloorNum(num: number, gridSizeNum: number): number {
  return Math.floor(num / gridSizeNum);
}

/**
 * 像素位置转网格位置
 */
export function px2gridPos(
  pos: Position,
  gridSize: Size,
  isMagnet: boolean = false
): Position {
  const calcFn = isMagnet ? calcPx2GridMagnetNum : calcPx2GridFloorNum;

  return {
    x: calcFn(pos.x, gridSize.width),
    y: calcFn(pos.y, gridSize.height),
  };
}

/**
 * 标准化Rect
 */
export function normalizeRect(rect: Rect): Rect {
  return {
    x1: Math.min(rect.x1, rect.x2),
    y1: Math.min(rect.y1, rect.y2),
    x2: Math.max(rect.x1, rect.x2),
    y2: Math.max(rect.y1, rect.y2),
  };
}

/**
 * 判断rect1是否在rect2中
 */
export function isRectInsideRect(rect1: Rect, rect2: Rect) {
  rect1 = normalizeRect(rect1);
  rect2 = normalizeRect(rect2);

  return (
    rect1.x1 >= rect2.x1 &&
    rect1.x2 <= rect2.x2 &&
    rect1.y1 >= rect2.y1 &&
    rect1.y2 <= rect2.y2
  );
}

/**
 * 将pxRect转化成gridRect
 * 算法是网里缩
 */
export function pxRect2GridRect(rect: Rect, gridSize: Size): Rect {
  rect = normalizeRect(rect);

  return {
    x1: Math.ceil(rect.x1 / gridSize.width),
    y1: Math.ceil(rect.y1 / gridSize.height),
    x2: Math.floor(rect.x2 / gridSize.width),
    y2: Math.floor(rect.y2 / gridSize.height),
  };
}

/**
 * 判断一个点是否在一个矩形空间内
 */
export function isPositionInsizeRect(pos: Position, rect: Rect): boolean {
  rect = normalizeRect(rect);

  return (
    pos.x <= rect.x2 && pos.x >= rect.x1 && pos.y <= rect.y2 && pos.y >= rect.y1
  );
}

/**
 * 根据宽高和中心点返回一个矩形
 */
export function getRectWithSize(pos: Position, size: Size): Rect {
  return {
    x1: pos.x - size.width / 2,
    y1: pos.y - size.height / 2,
    x2: pos.x + size.width / 2,
    y2: pos.y + size.height / 2,
  };
}
