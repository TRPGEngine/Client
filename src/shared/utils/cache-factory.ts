import { getStorage } from '@shared/manager/storage';
import _isNil from 'lodash/isNil';

export enum CachePolicy {
  /**
   * 存储在内存中
   */
  Temporary,

  /**
   * 存储在rnStorage中
   * 过期时间为rnStorage的过期时间
   * NOTE: 在使用持久化存储时必须自行实现hash
   * 可以基于buildCacheHashPrefix来快速实现
   */
  Persistent,
}

/**
 * 序列化参数
 */
function stringifyArgs(arg: any): string {
  switch (typeof arg) {
    case 'string':
      return arg;
    case 'number':
      return String(arg);
    default:
      return JSON.stringify(arg);
  }
}

/**
 * 拼接参数
 */
function joinArgs(args: any[]): string {
  return args.map(stringifyArgs).join('$');
}

function defaultHash(...args: any[]): string {
  return `cache:factory$${joinArgs(args)}`;
}

/**
 * 快速构建缓存的hash
 * @param prefix 前缀
 */
export function buildCacheHashFnPrefix(
  prefix: string
): (...args: any[]) => string {
  return (...args: any[]) => `cache:factory$${prefix}:${joinArgs(args)}`;
}

export function buildCacheFactory<T>(
  policy: CachePolicy,
  fn: (...args: any[]) => any,
  hashFn: (...args: any[]) => string = defaultHash
): (...args: any[]) => Promise<T> {
  let cacheManager: CacheManager;
  if (policy === CachePolicy.Temporary) {
    cacheManager = new TemporaryCache();
  } else if (policy === CachePolicy.Persistent) {
    cacheManager = new PersistentCache();
  }

  return async (...args: any[]) => {
    const hash = hashFn(...args);

    const cache = await cacheManager.get(hash);
    if (!_isNil(cache)) {
      return cache;
    }

    const ret = await fn(...args);
    await cacheManager.set(hash, ret);

    return ret;
  };
}

interface CacheManager {
  get(hash: string): Promise<any>;

  set(hash: string, value: any): Promise<void>;
}

class TemporaryCache implements CacheManager {
  temporaryCacheMap = {};

  async get(hash: string) {
    return this.temporaryCacheMap[hash];
  }
  async set(hash: string, value: any) {
    this.temporaryCacheMap[hash] = value;
  }
}

class PersistentCache implements CacheManager {
  async get(hash: string) {
    return await getStorage().get(hash);
  }
  async set(hash: string, value: any) {
    getStorage().set(hash, value);
  }
}
