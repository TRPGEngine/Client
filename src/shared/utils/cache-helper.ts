import {
  getUserInfo,
  getGroupInfo,
  getTemplateInfo,
  getActorInfo,
  getGroupInviteInfo,
  getFriendInviteInfo,
} from '@redux/actions/cache';
import _isNil from 'lodash/isNil';
import _get from 'lodash/get';
import { isUserUUID } from './uuid';
import type { CacheKey } from '@redux/types/cache';
import type { TRPGAction } from '@redux/types/__all__';

let _store;
export const attachStore = function (store) {
  _store = store;
};

/**
 * 异步请求用户信息缓存
 * 以确保缓存存在
 */
export function ensureUserInfoCache(uuid: string, type = 'user') {
  if (!uuid || typeof uuid !== 'string') {
    // 如果UUID为falsy 或 类型不为字符串。 则跳过
    return;
  }

  if (uuid.toString().substr(0, 4) === 'trpg') {
    // 不检测trpg开头的内置系统用户
    return;
  }

  if (!isUserUUID(uuid)) {
    console.warn('该UUID不是一个合法的UUID:', uuid);
    return;
  }

  const store = _store;
  if (!!store && !!store.dispatch) {
    if (type === 'user') {
      getUserInfoCache(uuid);
    }
  } else {
    throw new Error('checkUser func should bind store');
  }
}

/**
 * @deprecated 请使用 ensureUserInfoCache
 */
export const checkUser = function (uuid, type = 'user') {
  ensureUserInfoCache(uuid, type);
};

export const checkTemplate = function (uuid) {
  const store = _store;
  if (!!store && !!store.dispatch) {
    const state = store.getState();
    const info = _get(state, ['cache', 'template', uuid]);
    if (!info) {
      store.dispatch(getTemplateInfo(uuid));
    }
  } else {
    throw new Error('checkUser func should bind store');
  }
};

export const savecache = function () {
  // TODO
};

export const loadcache = function () {
  // TODO
};

type GetCacheCompletedCallback = () => void;
export type GetCacheDispatchActionFn = (
  uuid: string,
  onCompleted: GetCacheCompletedCallback
) => TRPGAction;

interface ReduxCacheFactoryInstance {
  (uuid: string): any;

  refresh: (uuid: string) => void;
}

/**
 * 一个工厂类，用于生成获取缓存的方法
 * 获取数据缓存，如果缓存已存在则从缓存中获取数据
 * example: const getUserInfoCache = reduxCacheFactory('user');
 * @param cacheScope 缓存对象, 必须是redux cache 下的变量
 */
function reduxCacheFactory(
  cacheScope: CacheKey,
  getCacheDispatch: GetCacheDispatchActionFn
): ReduxCacheFactoryInstance {
  const isGettingDataUUIDList: string[] = [];

  /**
   * 从服务端获取缓存信息
   */
  function _fetchCacheFromServer(uuid: string) {
    if (isGettingDataUUIDList.indexOf(uuid) === -1) {
      // 没有正在获取用户信息
      console.log(`缓存[${cacheScope}: ${uuid}]不存在， 自动获取`);
      _store.dispatch(
        getCacheDispatch(uuid, () => {
          // 从列表中移除
          const index = isGettingDataUUIDList.indexOf(uuid);
          if (index !== -1) {
            isGettingDataUUIDList.splice(index, 1);
          }
        })
      );
      isGettingDataUUIDList.push(uuid);
    }
  }

  function getCache(uuid: string) {
    if (_isNil(_store) || _isNil(_store.dispatch)) {
      throw new Error('get cache func should bind store');
    }

    if (_isNil(uuid) || uuid.toString().substr(0, 4) === 'trpg') {
      // 如果uuid为undefined或null
      // 或以trpg开头
      // 直接返回空Map
      return {};
    }

    const state = _store.getState();
    const data = _get(state, ['cache', cacheScope, uuid]);
    if (!_isNil(data)) {
      return data;
    }

    _fetchCacheFromServer(uuid);

    return {};
  }

  getCache.refresh = (uuid: string) => {
    // 重新获取缓存
    _fetchCacheFromServer(uuid);
  };

  return getCache;
}

/**
 * 用户信息缓存
 * 获取用户信息并缓存，如果缓存中已经有记录了则从缓存中获取
 * @param {string} uuid 用户UUID
 */
export const getUserInfoCache = reduxCacheFactory(
  'user',
  (uuid: string, onCompleted: GetCacheCompletedCallback) =>
    getUserInfo(uuid, onCompleted)
);

/**
 * 团信息缓存
 * 获取团信息并缓存，如果缓存中已经有记录了则从缓存中获取
 * @param {string} uuid 团UUID
 */
export const getGroupInfoCache = reduxCacheFactory(
  'group',
  (uuid: string, onCompleted: GetCacheCompletedCallback) =>
    getGroupInfo(uuid, onCompleted)
);

/**
 * 角色信息缓存
 * 获取角色信息并缓存，如果缓存中已经有记录了则从缓存中获取
 * @param {string} uuid 角色UUID
 */
export const getActorInfoCache = reduxCacheFactory(
  'actor',
  (uuid: string, onCompleted: GetCacheCompletedCallback) =>
    getActorInfo(uuid, onCompleted)
);

/**
 * 模板信息缓存
 * 获取模板信息并缓存，如果缓存中已经有记录了则从缓存中获取
 * @param {string} uuid 角色UUID
 */
export const getTemplateInfoCache = reduxCacheFactory(
  'template',
  (uuid: string, onCompleted: GetCacheCompletedCallback) =>
    getTemplateInfo(uuid, onCompleted)
);

/**
 * 好友邀请信息缓存
 * 获取信息并缓存，如果缓存中已经有记录了则从缓存中获取
 * @param {string} uuid 角色UUID
 */
export const getFriendInviteInfoCache = reduxCacheFactory(
  'friendInvite',
  (uuid: string, onCompleted: GetCacheCompletedCallback) =>
    getFriendInviteInfo(uuid, onCompleted)
);

/**
 * 团邀请信息缓存
 * 获取信息并缓存，如果缓存中已经有记录了则从缓存中获取
 * @param {string} uuid 角色UUID
 */
export const getGroupInviteInfoCache = reduxCacheFactory(
  'groupInvite',
  (uuid: string, onCompleted: GetCacheCompletedCallback) =>
    getGroupInviteInfo(uuid, onCompleted)
);

/**
 * 获取缓存的用户名
 * @param uuid 用户UUID
 */
export const getCachedUserName = (uuid: string): string => {
  const info = getUserInfoCache(uuid);
  return info.nickname ?? info.username ?? '';
};
