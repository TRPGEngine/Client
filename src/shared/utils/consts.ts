/**
 * 这里存储了一些常量
 * 用于在 rnStorage 或其他一些地方使用来保持一致性
 */

/**
 * 是否使用新应用
 */
export const IS_NEW_APP = '__isNewApp';

/**
 * 是否为开发人员
 * 用于处理一些仅开发人员可见的东西
 */
export const IS_DEVELOPER = '__isDeveloper';

/**
 * 是否为内测用户
 */
export const IS_ALPHA_USER = 'isAlphaUser';

/**
 * 用于打印RTC调试的标识
 */
export const RTC_DEBUG = '__rtc_debug';

/**
 * web端
 * 是否关闭了appBanner
 * 用于确保如果用户手动关闭其不多次显示
 */
export const APP_BANNER_CLOSE = 'appBannerClose';

/**
 * 公告是否关闭
 */
export const ANNOUNCEMENT_BAR_CLOSE = 'announcementBarClose';

/**
 * 系统消息用的UUID
 */
export const SYSTE_CONVERSE_SPEC = 'trpgsystem';

/**
 * 系统语言的常量
 */
export const LANGUAGE_KEY = 'trpg:i18n:language';

/**
 * portal端的jwt存储key
 */
export const PORTAL_JWT_KEY = 'jwt';

/**
 * 小程序端的jwt存储key
 */
export const TARO_JWT_KEY = 'jwt';

/**
 * 是否已安装PWA
 * 仅用于web端 chrome浏览器
 */
export const PWA_INSTALLED = 'pwaInstalled';

/**
 * 可以手动指定远程服务器的地址, 用于测试
 */
export const SERVICE_URL = 'serviceUrl';
