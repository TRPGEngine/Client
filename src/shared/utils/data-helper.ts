import type { GroupActorType } from '@src/shared/redux/types/group';
import _get from 'lodash/get';
import _isNil from 'lodash/isNil';
import type { PlayerUser } from '@shared/model/player';
import type { MsgPayload } from '@redux/types/chat';

/**
 * 数据相关帮助函数
 */

/**
 * 获取团人物卡字段
 * @param groupActor 团人物卡
 * @param field 字段名
 */
export const getGroupActorField = (
  groupActor: GroupActorType,
  field: 'avatar' | 'name' | 'desc'
): string => {
  return groupActor[field] || _get(groupActor, ['actor', field]);
};
export const getGroupActorName = (groupActor: GroupActorType) =>
  getGroupActorField(groupActor, 'name');

/**
 * 获取团人物卡信息
 * 为在原始团人物信息的基础上复写团人物信息
 * @param groupActor 团人物卡
 */
export const getGroupActorInfo = (groupActor: GroupActorType): {} => {
  return {
    ..._get(groupActor, 'actor.info', {}),
    ...groupActor.actor_info,
  };
};

/**
 * 获取团人物卡的模板UUID
 * @param groupActor 团人物卡
 */
export const getGroupActorTemplateUUID = (
  groupActor: GroupActorType
): string => {
  return (
    groupActor.actor_template_uuid ?? _get(groupActor, 'actor.template_uuid')
  );
};

/**
 * 获取用户名
 * @param userInfo 用户信息
 */
export function getUserName(
  userInfo: Partial<PlayerUser> | null | undefined
): string {
  if (_isNil(userInfo)) {
    return '';
  }

  return userInfo?.nickname || userInfo?.username || '';
}

export function buildMsgPayload(payload: Partial<MsgPayload>): MsgPayload {
  return {
    type: 'normal',
    uuid: '',
    sender_uuid: '',
    message: '',
    is_public: false,
    is_group: false,
    date: new Date().toISOString(),
    revoke: false,
    ...payload,
  };
}
