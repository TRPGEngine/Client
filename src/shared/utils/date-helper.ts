import moment, { MomentInput } from 'moment';
import './date-init';

/**
 * 日期相关帮助函数
 */

/**
 * 获取时间距离现在的可读化结果
 * @param str 时间
 */
export function getFromNow(str?: MomentInput): string {
  const date = str ? moment(str) : moment();
  return date.fromNow();
}
export const getShortDiff = (str: MomentInput): string => {
  const date = str ? moment(str) : moment();
  if (helper.isToday(date)) {
    return date.format('HH:mm');
  } else {
    return date.fromNow();
  }
};
export const getShortDate = (str?: string): string => {
  const date = str ? moment(str) : moment();
  if (helper.isToday(date)) {
    return date.format('A hh:mm');
  } else {
    return date.format('MM-DD HH:mm');
  }
};
export const getMsgDate = (str: string): string => {
  const date = str ? moment(str) : moment();
  if (helper.isToday(date)) {
    return date.format('HH:mm:ss');
  } else {
    return date.format('YYYY-MM-DD HH:mm:ss');
  }
};
export const getFullDate = (str?: string): string => {
  const date = str ? moment(str) : moment();
  return date.format('YYYY-MM-DD HH:mm:ss');
};
export const getSimpleDate = (str: string): string => {
  return str ? moment(str).format('YYYY-MM-DD') : '';
};
export const isToday = (date: MomentInput): boolean => {
  return moment().isSame(date, 'day');
};
export const getDateDiff = (
  prevDate: MomentInput,
  nextDate: MomentInput
): number => {
  return moment(nextDate).valueOf() - moment(prevDate).valueOf();
};

/**
 * 判定是否为同一天
 */
export function isSameDate(
  prevDate: MomentInput,
  nextDate: MomentInput
): boolean {
  return (
    moment(prevDate).format('YYYY-MM-DD') ===
    moment(nextDate).format('YYYY-MM-DD')
  );
}

/**
 * 判断是否需要强调时间(两个日期间隔超过设定值)
 */
export const shouleEmphasizeTime = (
  prevDate: MomentInput,
  nextDate: MomentInput
): boolean => {
  const diffTime = getDateDiff(prevDate, nextDate);
  return diffTime / 1000 / 60 >= 5; // 超过5分钟
};

const helper = {
  getShortDiff,
  getShortDate,
  getMsgDate,
  getFullDate,
  getSimpleDate,
  isToday,
  getDateDiff,
};

export default helper;
