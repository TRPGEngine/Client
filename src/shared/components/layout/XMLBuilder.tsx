import React, {
  useMemo,
  useState,
  useEffect,
  useCallback,
  useImperativeHandle,
} from 'react';
import parser, { XMLElement } from './parser/xml-parser';
import * as processor from './processor';
import _isEmpty from 'lodash/isEmpty';
import _isUndefined from 'lodash/isUndefined';
import _isNil from 'lodash/isNil';
import _isFunction from 'lodash/isFunction';
import './tags/__all__';
import styled from 'styled-components';
import { StateDataType, StateActionType, preRenderTags } from './types';
import { useSize } from 'react-use';
import { LayoutWidthContextProvider } from './context/LayoutWidthContext';
import { LayoutStateContextProvider } from './context/LayoutStateContext';
import { useBuildLayoutStateContext } from './hooks/useBuildLayoutStateContext';
import { iterativeConfigKey } from './parser/key';
import { TMemo } from '../TMemo';
import { LayoutNode } from './processor/LayoutNode';
import {
  getASTObjectWithoutTagName,
  getASTObjectWithTagName,
} from './parser/tagname-helper';

/**
 * XML布局需求:
 * name必须全局唯一
 * 需要版本号
 * 以Template组件作为统一包装(不强求)
 */

export type DefinePropsType = {
  [name: string]: any;
};
interface DefineMap {
  [name: string]: React.ComponentType<DefinePropsType>;
}

export interface DataMap {
  _name?: string;
  _avatar?: string;
  _desc?: string;
  [name: string]: StateDataType | DataMap; // 可以嵌套
}

export type LayoutType = 'edit' | 'detail';

export interface XMLBuilderState {
  defines: DefineMap;
  data: DataMap;
  global: DataMap; // 全局变量
}

export interface XMLBuilderAction {
  type: StateActionType;
  payload: { [name: string]: any };
  [others: string]: any;
}

export interface XMLBuilderContext {
  state: XMLBuilderState;
  dispatch: React.Dispatch<XMLBuilderAction>;
  layoutType: LayoutType;
  parent?: XMLBuilderContext;
}

export type StateChangeHandler = (newState: XMLBuilderState) => void;

const XMLBuilderContainer = styled.div`
  text-align: left;
  width: 100%;

  pre {
    margin: 0;
  }

  code {
    padding: 0.2em 0.4em;
    margin: 0;
    font-size: 85%;
    background-color: rgba(27, 31, 35, 0.05);
    border-radius: 3px;
  }
`;

export class XMLErrorBoundary extends React.Component {
  state = { hasError: false, error: null };

  static getDerivedStateFromError(error) {
    // 更新 state 使下一次渲染能够显示降级后的 UI
    return { hasError: true, error };
  }

  componentDidCatch(error, errorInfo) {
    // 你同样可以将错误日志上报给服务器
    console.error(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      // 你可以自定义降级后的 UI 并渲染
      return (
        <div>
          <h2>人物卡布局解析错误</h2>
          <p>{String(this.state.error)}</p>
        </div>
      );
    }

    return this.props.children;
  }
}

/**
 * 容器与上下文抽象
 */
export const LayoutProviderContainer: React.FC<{
  style?: React.CSSProperties;
  layoutType?: LayoutType;
  initialData?: DataMap;
  onChange?: StateChangeHandler;
  builderRef: React.Ref<XMLBuilderRef>;
}> = TMemo((props) => {
  const { style, layoutType = 'edit', builderRef } = props;
  const handleChange = useCallback(
    (newState: XMLBuilderState) => {
      _isFunction(props.onChange) && props.onChange(newState);
    },
    [props.onChange]
  );
  const { state, dispatch } = useBuildLayoutStateContext({
    initialData: props.initialData ?? {},
    onChange: handleChange,
  });

  const context = useMemo(() => ({ state, dispatch, layoutType }), [
    state,
    dispatch,
    layoutType,
  ]);

  const [XMLRender, { width }] = useSize(
    <XMLBuilderContainer style={style}>{props.children}</XMLBuilderContainer>
  );

  useImperativeHandle(
    builderRef,
    () => ({
      updateAllData: (data) => {
        dispatch({
          type: StateActionType.UpdateAllData,
          payload: {
            data,
          },
        });
      },
    }),
    []
  );

  return (
    <LayoutWidthContextProvider width={width}>
      <LayoutStateContextProvider state={context}>
        {XMLRender}
      </LayoutStateContextProvider>
    </LayoutWidthContextProvider>
  );
});
LayoutProviderContainer.displayName = 'LayoutProvider';

interface XMLBuilderProps {
  xml: string;
  layoutType?: LayoutType;
  initialData?: DataMap;
  onChange?: StateChangeHandler;
}
export interface XMLBuilderRef {
  /**
   * 更新所有数据
   */
  updateAllData: (data: Record<string, any>) => void;
}
export const XMLBuilder = TMemo(
  React.forwardRef<XMLBuilderRef, XMLBuilderProps>((props, ref) => {
    const { xml = '', onChange, layoutType = 'edit' } = props;
    const [error, setError] = useState<Error | null>(null);
    const [layout, setLayout] = useState<XMLElement>();

    useEffect(() => {
      try {
        const ast = parser(xml);
        ast.type = 'root';
        iterativeConfigKey(ast);
        console.log('layout', ast);

        // 渲染分两步
        // 1. 渲染预加载组件
        setLayout(getASTObjectWithTagName(ast, preRenderTags));
        requestAnimationFrame(() => {
          // 2. 待渲染完毕后渲染除了预加载组件以外的所有组件
          setLayout(getASTObjectWithoutTagName(ast, preRenderTags));
        });

        setError(null);
      } catch (err) {
        console.error(err);
        setError(err);
      }
    }, [xml]);

    const LayoutDOM = useMemo(() => {
      if (!_isNil(error)) {
        return (
          <div>
            <p>XML解析出现错误:</p>
            <p>{String(error)}</p>
          </div>
        );
      }

      if (_isEmpty(layout)) {
        return null;
      }

      try {
        return <LayoutNode node={layout!} />;
      } catch (err) {
        console.error(err);
        setError(err);
        return null;
      }
    }, [layout, error]);

    return (
      <XMLErrorBoundary>
        <LayoutProviderContainer builderRef={ref}>
          {LayoutDOM}
        </LayoutProviderContainer>
      </XMLErrorBoundary>
    );
  })
);
XMLBuilder.displayName = 'XMLBuilder';

export default XMLBuilder;
