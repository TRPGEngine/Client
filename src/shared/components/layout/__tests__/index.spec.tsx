import React from 'react';
import XMLBuilder from '@shared/components/layout/XMLBuilder';
import { render } from '@testing-library/react';
import { sleep } from '@test/utils';

const wrapXML = (xml: string) => `<Template>${xml}</Template>`;
const renderXML = (xml: string) => render(<XMLBuilder xml={wrapXML(xml)} />);

describe('layout', () => {
  test('basic', async () => {
    const wrapper = renderXML('');
    expect(wrapper.asFragment()).toMatchSnapshot();

    await sleep(100);

    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  test('some text', async () => {
    const text = '<span>some text</span>';
    const wrapper = renderXML(text);

    expect(wrapper.asFragment()).toMatchSnapshot();

    await sleep(100);

    expect(wrapper.asFragment()).toMatchSnapshot();
    expect(wrapper.getByText('some text')).toBeInTheDocument();
  });
});
