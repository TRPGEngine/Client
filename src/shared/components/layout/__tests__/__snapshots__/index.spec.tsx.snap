// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout basic 1`] = `
<div
  className="sc-fzpans jRldgn"
  style={
    Object {
      "position": "relative",
    }
  }
>
  <iframe
    style={
      Object {
        "background": "transparent",
        "border": "none",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
        "zIndex": -1,
      }
    }
  />
</div>
`;

exports[`layout some text 1`] = `
<div
  className="sc-fzpans jRldgn"
  style={
    Object {
      "position": "relative",
    }
  }
>
  <iframe
    style={
      Object {
        "background": "transparent",
        "border": "none",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
        "zIndex": -1,
      }
    }
  />
  <div
    className="sc-AxgMl hoELXQ"
  >
    <span>
      some text
    </span>
  </div>
</div>
`;
